File: BInary/SmallestDivisor.cpp
// Problem: divide all number with any number and take ceil of that number and return the sum of the all number.
// Return smallest divisor which is less than or equal to limit

// Find sum of all number
int smallDiv(vector<int> &arr, int mid)
{
    int ans = 0;
    for (int i = 0; i < arr.size(); i++)
    {
        if (arr[i] % mid == 0)
        {
            ans += arr[i] / mid;
        }
        else
        {
            ans += (arr[i] / mid) + 1;
        }
    }
    return ans;
}
int smallestDivisor(vector<int> &arr, int limit)
{
    // Write your code here.
    int n = arr.size();

    if (n > limit)
        return -1;

    int low = 1;
    int high = *max_element(arr.begin(), arr.end());

    for (int i = 0; i < arr.size(); i++)
    {
        high = max(high, arr[i]);
    }

    while (low <= high)
    {
        int mid = (low + high) / 2;
        int ans = smallDiv(arr, mid);
        // If number is found then we don't need to move upword direction so we take high = mid - 1
        // at the end of the day low is pointing your smallest divisor.
        if (ans <= limit)
        {
            high = mid - 1;
        }
        else
        {
            low = mid + 1;
        }
    }
    return low;
}

File: Binary_Search/SmallestDivisor.java
class Solution {
    public int findMaxEl(int[] arr, int n) {
        int maxEl = Integer.MIN_VALUE;
        for(int i = 0; i < n; i++) {
            maxEl = Math.max(maxEl, arr[i]);
        }
        return maxEl;
    }

    public boolean findThreshold(int[] arr, int divisor, int threshold) {
        int sum = 0;
        for(int i = 0; i < arr.length; i++) {
            if(arr[i] <= divisor) {
                sum++;
            } else if((arr[i] % divisor) == 0){
                sum += (arr[i] / divisor);
            } else {
                sum += (arr[i] / divisor) + 1;
            }
            
        }
        if(sum <= threshold) {
            return true;
        }

        return false;
    }
    
    public int smallestDivisor(int[] nums, int threshold) {
        int n = nums.length;
        if(threshold < n) {
            return -1;
        }

        int maxEl = findMaxEl(nums, n);
        int low = 1;
        int high = maxEl;

        if(threshold == n) {
            return maxEl;
        }

        int ans = -1;
        while(low <= high) {
            int mid = (low + high) / 2;
            boolean isValid = findThreshold(nums, mid, threshold);
            if(isValid) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return low;
        // return ans;
    }
}